Adjustable Container

Edru2 opened the pull request:
#### Brief overview of PR changes/additions
Adjustable Containers are a new Geyser Class which can act as Geyser Containers with some extras like: moving/resizing by mouse, saving and loading of adjustments, attaching element to a border and so on.

Many of these functions are available by a menu which opens by right mouse click at the top border of the container.

A  (unfinished for now) documentation can be found here: https://github.com/Edru2/AdjustableContainer/wiki

A small video to show how they can be used is here: https://www.youtube.com/watch?v=GT2ScizuM48&feature=youtu.be

#### Motivation for adding to Mudlet
They are very flexible and allow to create nice GUIs easily.

#### Other info (issues closed, discussion etc)
Will finish the documentation in the next days.

Some new functions in this version (undocumented for now) are:

**Adjustable.Container.showAll()** --shows all hidden containers

**mycontainer:newCustomItem**("ItemName", function() echo("this is a test") end) 
-- creates a new menu "custom" with the new Item "ItemName" which will execute the function if clicked on

It's also possible to use the own container as parameter:
**mycontainer:newCustomItem**("MoveToZero", function(self) self:move(0,0) end)

A new function to do any action on all Adjustable Containers
**Adjustable.Container.doAll**(function(self) self:hide() end) --hides all Adjustable.Containers

**mycontainer:setAbsolute(bool size, bool position)** --can set the size or the position as absolute value. Normally all Adjustable.Containers set all values in % values.
...

Edru2 committed:
initial commit

Adjustable Container.
Geyser Container Class with a few additions

vadi2 commented on src/mudlet-lua/lua/geyser/GeyserAdjustableContainer.lua:
```diff
@@ -0,0 +1,736 @@
+
+Adjustable = Adjustable or {}
+Adjustable.Container = Adjustable.Container or Geyser.Container:new({
+    name = "AdjustableContainerClass",})
```
```suggestion
    name = "AdjustableContainerClass"})
```

Edru2 committed:
Make menu, button, font size adjustable

(Almost) Everything should be adjustable

Edru2 committed:
reordering and reindenting

Edru2 committed:
lockstyle works now as expected, some polishing, added raiseOnClick

Edru2 committed:
unlockContainer fix order

Edru2 committed:
Comments update/ Color emojis in Linux also in userwindows

Wrote comment for every function in order to work with ldoc.
In Linux emoji of the menu items were not colored if the container was located in an userwindow

Edru2 commented:
I made a new version of the GUIDropManager (Drag and Drop Images onto Mudlet) which needs Adjustable Container to work with.
Download link: [GUIDropManager.zip](https://github.com/Mudlet/Mudlet/files/4473462/GUIDropManager.zip)
Link to the comment: https://github.com/Mudlet/Mudlet/issues/1399#issuecomment-601639527
Link to a small youtube video: https://youtu.be/Zl7h83SA9bY

vadi2 commented:
Very nice! Ask for testers on this too.

vadi2 commented on src/mudlet-lua/lua/geyser/GeyserAdjustableContainer.lua:
```diff
@@ -0,0 +1,833 @@
+    
+    if event.button ~= "LeftButton" and not self.minimized then
+        if (top and not left and not right) or (bottom and not left and not right) then
+            label:setCursor("SizeVer")
```
Maybe we can improve on the naming here - this doesn't read all that simple

Edru2 replied:
I'm not sure how the naming could be better but I cleaned up a bit the if statements.
It should be better readable now.

vadi2 replied:
Oh I meant the cursor. What size "size ver" mean? Not much without looking it up...

vadi2 replied:
That's not for this PR, I should have clarified. Just looking at how the new cursors look in real code :)

Edru2 replied:
Oh that's from CursorShapes directly from the qt documentation here https://doc.qt.io/qt-5/qt.html#CursorShape-enum. Has nothing to do with Adjustable Containers.

vadi2 replied:
Yeah... but at the end of the day we're responsible for how things are, can't point the finger at someone else. Anyway yes, nothing to do with this PR!

Edru2 replied:
Happy to add better names for CursorShapes. If someone more creative and with better language skills then me makes some suggestions :smiley:

vadi2 replied:
Sure, I've thought of them here: https://github.com/Mudlet/Mudlet/issues/3607

vadi2 commented on src/mudlet-lua/lua/geyser/GeyserAdjustableContainer.lua:
```diff
@@ -0,0 +1,833 @@
+end
+
+--- internal function to check which valid attach position the container is at
+function Adjustable.Container:fPos()
```
What does "fPos" mean - could it have a better name?

vadi2 commented on src/mudlet-lua/lua/geyser/GeyserAdjustableContainer.lua:
```diff
@@ -0,0 +1,833 @@
+function Adjustable.Container:fPos()
+    local winw, winh = getMainWindowSize()
+    local fpos = {}
+    if  (winh*0.8)-self.get_height()<= self.get_y()  then  table.insert(fpos,"bottom") end
```
```suggestion
    if  (winh*0.8)-self.get_height()<= self.get_y()  then fpos[#fpos+1] = "bottom" end
```

This is quicker, change others too

vadi2 commented on src/mudlet-lua/lua/geyser/GeyserAdjustableContainer.lua:
```diff
@@ -0,0 +1,833 @@
+    mytable.padding = self.padding
+    mytable.hidden = self.hidden
+    mytable.auto_hidden = self.auto_hidden
+    if not(io.exists(getMudletHomeDir().."/AdjContainer/")) then lfs.mkdir(getMudletHomeDir().."/AdjContainer/") end
```
Could it have a full name? We don't have a limit on characters ;)

Edru2 committed:
added suggestions

vadi2 reviewed (approved):
Great work, let's try it out! Could you add documentation for it?

Edru2 commented:
> Great work, let's try it out! Could you add documentation for it?

The documentation can be found here: https://github.com/Edru2/AdjustableContainer/wiki
After it's merged I will add it to the Mudlet wiki.

vadi2 commented:
It's good to be merged now, and it'll be in tonight's PTB then without documentation - unless we point people to the old location, but then it'll be confusing when we change it. Easier to point to a single, final location.

As a workaround you could even stub documentation in the wiki, saying it'll be migrated soon but for now look there. Just so people start remembering a consistent location go to / to link others to

vadi2 commented:
It's in! celebration time 😎😎

Leigheas commented:
Ok couple things I've noticed starting to use the Adjustable Containers:

1: its not an always happening thing, but sometimes when I unlock a container it will stretch to the whole height of the screen.

2: also I will lock a container into place (thru the right click), save it, and then close mudlet (making sure to save my profile). when I reopen the profile, all the windows are back into their default positions.
 I have also tried Adjustable.Container.saveAll() and that didn't save them as well.

3:  Its drawing one of the windows twice and only does it to that window, (my code redraws the ui once it recieves some data as to populate the correct values.) which in turn makes it act extremely wonky with anything dealing with that window.

Edru2 commented:
Thanks for feedback!
1: Is there a way I can reproduce that? I mean some example code. It could probably be the same issue as mentioned in 3

2: Did you load the container at start with container:load() or Adjustable.Container:loadAll()

3: Hmm do you use code like **mycontainer = mycontainer or Adjustable.Container:new(...)** the problem is that if you don't write it like this every reload of the ui makes duplicates (I think it's also the same for normal Geyser.Container but there you wouldn't notice it). Thinking about a fix for that, but for now you should write the code as above.

Leigheas commented:
> Thanks for feedback!
> 1: Is there a way I can reproduce that? I mean some example code. It could probably be the same issue as mentioned in 3
> 
I'm thinking this perhaps is more on my side (probably my computer being funny since it hasn't happened since), will do more testing.
> 2: Did you load the container at start with container:load() or Adjustable.Container:loadAll()
> 
Actually I was still loading the windows thru the normal geyser method (I completely overlooked the loadAll() container function, my apologies

> 3: Hmm do you use code like **mycontainer = mycontainer or Adjustable.Container:new(...)** the problem is that if you don't write it like this every reload of the ui makes duplicates (I think it's also the same for normal Geyser.Container but there you wouldn't notice it). Thinking about a fix for that, but for now you should write the code as above.
I double checked the one container that was replicating and I did notice I forgot the mycontainer or part so it was making them on each draw..

Leigheas commented:
just an update, I noticed for some reason I have to have the containers created and visible then I can do loadAll(). 
What I did was create the containers, saved all them, then exited the profile making sure to save them. When I would go back into the profile and log into my mud, I have it in the code on a connection event to do the loadAll() and showAll(). The boxes would not show up.
So I ran the function I made that created all them again, then did loadAll() and it put them back into their saved positions

I was under the impression that loadAll() would load the containers and put them in their saved spots.

Edru2 commented:
loadAll/showAll cannot work if the container are not created (there is nothing to show/load). I suggest to put the your_container:load() or loadAll directly after creation of your container.

You gave me idea to directly integrate load (and maybe also save) into the container creation (as a constraint). 
I'm not sure if I will be able to integrate that before the next release but it is on my list.
For now the best would be to put load directly after the container creation and as an Idea maybe put saveAll() handled by sysExitEvent.

Leigheas commented:
> loadAll/showAll cannot work if the container are not created (there is nothing to show/load). I suggest to put the your_container:load() or loadAll directly after creation of your container.
> 
> You gave me idea to directly integrate load (and maybe also save) into the container creation (as a constraint).
> I'm not sure if I will be able to integrate that before the next release but it is on my list.
> For now the best would be to put load directly after the container creation and as an Idea maybe put saveAll() handled by sysExitEvent.

yeah thats how I figured it out.
I have it create the containers, then loadAll().
and on the sysExitEvent it does a saveAll()

vadi2 commented:
Could you add that to the manual?

Leigheas commented:
> Could you add that to the manual?

I did a small note stating you have to load the containers first, sorry not used to wiki editing
and if it makes more logic to be in another place then I apologize

vadi2 commented:
It's cool, thanks for adding :+1:
